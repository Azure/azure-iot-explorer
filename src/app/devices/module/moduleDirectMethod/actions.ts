/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { DEVICECONTENT } from '../../../constants/actionPrefixes';
import { INVOKE_MODULE_IDENTITY_METHOD } from '../../../constants/actionTypes';

export const invokeModuleDirectMethodAction = actionCreatorFactory(DEVICECONTENT).async<InvokeModuleMethodActionParameters, string>(INVOKE_MODULE_IDENTITY_METHOD);

export interface InvokeModuleMethodActionParameters {
    connectTimeoutInSeconds: number;
    deviceId: string;
    methodName: string;
    moduleId: string;
    payload?: any; // tslint:disable-line:no-any
    responseTimeoutInSeconds: number;
}
