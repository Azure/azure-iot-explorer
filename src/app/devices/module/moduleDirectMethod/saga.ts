/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { call, put, takeEvery } from 'redux-saga/effects';
import { SagaIterator } from 'redux-saga';
import { Action } from 'typescript-fsa';
import { invokeModuleDirectMethodAction, InvokeModuleMethodActionParameters } from './actions';
import { invokeModuleDirectMethod } from '../../../api/services/moduleService';
import { raiseNotificationToast } from '../../../notifications/components/notificationToast';
import { NotificationType } from '../../../api/models/notification';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { InvokeModuleMethodParameters } from '../../../api/parameters/moduleParameters';

export function* invokeModuleDirectMethodSagaWorker(action: Action<InvokeModuleMethodActionParameters>): SagaIterator {
    const toastId: number = Math.random();

    try {
        const invokeMethodParameters: InvokeModuleMethodParameters = {
            ...action.payload,
        };

        yield call(notifyModuleMethodInvokedHelper, toastId, invokeMethodParameters);

        const response = yield call(invokeModuleDirectMethod, invokeMethodParameters);
        const stringifiedResponse = typeof response === 'object' ? JSON.stringify(response) : response;

        yield call(raiseNotificationToast, {
            id: toastId,
            text: {
                translationKey: ResourceKeys.notifications.invokeModuleMethodOnSuccess,
                translationOptions: {
                    deviceId: action.payload.deviceId,
                    methodName: action.payload.methodName,
                    moduleId: action.payload.moduleId,
                    response: stringifiedResponse
                },
            },
            type: NotificationType.success
        });

        yield put(invokeModuleDirectMethodAction.done({params: action.payload, result: stringifiedResponse}));
    } catch (error) {
        yield call(raiseNotificationToast, {
            id: toastId,
            text: {
                translationKey: ResourceKeys.notifications.invokeModuleMethodOnError,
                translationOptions: {
                    deviceId: action.payload.deviceId,
                    error,
                    moduleId: action.payload.moduleId
                },
            },
            type: NotificationType.error
        });

        yield put(invokeModuleDirectMethodAction.failed({params: action.payload, error}));
    }
}

export function* notifyModuleMethodInvokedHelper(toastId: number, payload: InvokeModuleMethodParameters) {
    if (payload) {
        yield call(raiseNotificationToast, {
            id: toastId,
            text: {
                translationKey: ResourceKeys.notifications.invokingModuleMethodWithPayload,
                translationOptions: {
                    deviceId: payload.deviceId,
                    methodName: payload.methodName,
                    moduleId: payload.moduleId,
                    payload: JSON.stringify(payload.payload),
                },
            },
            type: NotificationType.info
        });
        return payload.payload;
    }
}

export function* invokeModuleDirectMethodSaga() {
    yield takeEvery(invokeModuleDirectMethodAction.started.type, invokeModuleDirectMethodSagaWorker);
}
