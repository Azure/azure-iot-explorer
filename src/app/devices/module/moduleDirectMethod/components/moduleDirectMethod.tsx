/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useHistory } from 'react-router-dom';
import { CommandBar } from '@fluentui/react';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { DIRECT_METHOD, NAVIGATE_BACK } from '../../../../constants/iconNames';
import { useAsyncSagaReducer } from '../../../../shared/hooks/useAsyncSagaReducer';
import { invokeModuleDirectMethodSaga } from '../saga';
import { invokeModuleDirectMethodAction } from '../actions';
import { DirectMethodForm } from '../../../../devices/directMethod/components/directMethodForm';
import { ROUTE_PARAMS, ROUTE_PARTS } from '../../../../constants/routes';
import '../../../../css/_deviceDetail.scss';
import { AppInsightsClient } from '../../../../shared/appTelemetry/appInsightsClient';
import { TELEMETRY_PAGE_NAMES, TELEMETRY_USER_ACTIONS } from '../../../../../app/constants/telemetry';

const DEFAULT_TIMEOUT = 10;

export const ModuleDirectMethod: React.FC = () => {
    const { t } = useTranslation();
    const { search, pathname } = useLocation();
    const history = useHistory();
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);

    const [ , dispatch ] = useAsyncSagaReducer(() => undefined, invokeModuleDirectMethodSaga, undefined);
    const [connectionTimeOut, setConnectionTimeOut] = React.useState<number>(DEFAULT_TIMEOUT);
    const [methodName, setMethodName] = React.useState<string>('');
    const [payload, setPayload] = React.useState<string>('');
    const [responseTimeOut, setResponseTimeOut] = React.useState<number>(DEFAULT_TIMEOUT);

    const navigateToModuleList = () => {
        const path = pathname.replace(/\/moduleIdentity\/moduleMethod\/.*/, `/${ROUTE_PARTS.MODULE_IDENTITY}`);
        history.push(`${path}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`);
    };

    const showCommandBar = () => {
        return (
            <CommandBar
                className="command"
                items={[
                    {
                        ariaLabel: t(ResourceKeys.directMethod.invokeMethodButtonText),
                        disabled: !formReady(),
                        iconProps: {iconName: DIRECT_METHOD},
                        key: DIRECT_METHOD,
                        name: t(ResourceKeys.directMethod.invokeMethodButtonText),
                        onClick: onInvokeMethodClickHandler
                    }
                ]}
            />
        );
    };

    const formReady = (): boolean => !!methodName;
    const getPayload = (content: string) => {
        // payload could be json or simply string
        try {
            return JSON.parse(content);
        } catch {
            return content;
        }
    };

    const onInvokeMethodClickHandler = () => {
        AppInsightsClient.trackUserAction(TELEMETRY_USER_ACTIONS.MODULE_INVOKE_DIRECT_METHOD);
        dispatch(invokeModuleDirectMethodAction.started({
            connectTimeoutInSeconds: connectionTimeOut,
            deviceId,
            methodName,
            moduleId,
            payload: getPayload(payload),
            responseTimeoutInSeconds: responseTimeOut
        }));
    };

    React.useEffect(() => {
        AppInsightsClient.getInstance()?.trackPageView({name: TELEMETRY_PAGE_NAMES.MODULE_DIRECT_METHOD});
    }, []); // tslint:disable-line: align

    return (
        <>
            {showCommandBar()}
            <DirectMethodForm
                methodName={methodName}
                connectionTimeOut={connectionTimeOut}
                responseTimeOut={responseTimeOut}
                setMethodName={setMethodName}
                setPayload={setPayload}
                setConnectionTimeOut={setConnectionTimeOut}
                setResponseTimeOut={setResponseTimeOut}
            />
        </>
    );
};
