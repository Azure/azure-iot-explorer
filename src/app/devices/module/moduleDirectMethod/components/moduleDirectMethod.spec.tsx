/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';
import { ModuleDirectMethod } from './moduleDirectMethod';

const pathname = '#/devices/deviceDetail/moduleIdentity/moduleTwin/?deviceId=newdevice&moduleId=moduleId';
jest.mock('react-router-dom', () => ({
    useHistory: () => ({ push: jest.fn() }),
    useLocation: () => ({ search: '?deviceId=newdevice&moduleId=moduleId', pathname }),
}));

describe('moduleDirectMethod', () => {
    it('matches snapshot', () => {
        expect(shallow(<ModuleDirectMethod/>)).toMatchSnapshot();
    });
});
