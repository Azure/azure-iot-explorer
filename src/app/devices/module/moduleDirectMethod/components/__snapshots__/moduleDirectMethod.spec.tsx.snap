// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moduleDirectMethod matches snapshot 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    items={
      Array [
        Object {
          "ariaLabel": "directMethod.invokeMethodButtonText",
          "disabled": true,
          "iconProps": Object {
            "iconName": "Remote",
          },
          "key": "Remote",
          "name": "directMethod.invokeMethodButtonText",
          "onClick": [Function],
        },
      ]
    }
  />
  <DirectMethodForm
    connectionTimeOut={10}
    methodName=""
    responseTimeOut={10}
    setConnectionTimeOut={[Function]}
    setMethodName={[Function]}
    setPayload={[Function]}
    setResponseTimeOut={[Function]}
  />
</Fragment>
`;
